packages/pnpm/spnpm.js

```diff
@@ -1,3 +1,60 @@
-require('@pnpm/ts-execution-runtime')
+const fs = require('fs')
+const esbuild = require('esbuild')
+const pathLib = require('path')
+const packagesDir = pathLib.resolve(__dirname, '..')
 
-require('./src/pnpm.ts')
+const localPackages = fs.readdirSync(packagesDir)
+
+const pnpmPackageJson = JSON.parse(fs.readFileSync(pathLib.join(__dirname, 'package.json'), 'utf8'))
+
+// This plugin rewrites imports to reference the `src` dir instead of `lib` so
+// esbuild can compile the original TypeScript
+const spnpmImportsPlugin = {
+  name: 'spnpmImports',
+  setup: (build) => {
+    // This is an exception to the rule that all local packages start with `@pnpm`
+    build.onResolve({ filter: /^dependency-path$/ }, ({path, resolveDir}) => ({
+      path: pathLib.resolve(packagesDir, 'dependency-path', 'src', 'index.ts')
+    }))
+
+    // E.g. @pnpm/config -> /<some_dir>/pnpm/packages/config/src/index.ts
+    build.onResolve({ filter: /@pnpm\// }, ({path, resolveDir}) => {
+      const pathParts = path.split('/')
+      const packageName = pathParts[1]
+
+      // Bail if the package isn't present locally
+      if (!localPackages.includes(packageName)) {
+        return
+      }
+
+      const newPath = pathLib.resolve(packagesDir, packageName, 'src', 'index.ts')
```

Comment:
maybe better to use info from the lockfile to find the main file.

Reply:
Where specifically are you thinking? The `main` field in the `package.json`? I could read that in and `.replace('lib', 'src')` but that doesn't seem much better than simply hard coding it.

The other approach I could take is not writing an esbuild plugin at all. Instead I could write a secondary "main" field into each of the `packages/**/package.json` files e.g. `"tsMain": "src/index.ts"` and then invoke esbuild with `--mainField=tsMain`. The downside is that puts a mysterious `tsMain` field in every package.json.

Reply:
Yes, I mean the main field of the package.json files.

The seconds approach sounds fine as well.

Or keep it as it is if it works. I don't care that much.